RavenDB-13106 first steps towards resharding

karmeli87 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-13106 

### Additional description

First PR of many to come. 

First and still draft implementation of :
- Add relevant fields to the database record
- New migration replication type 
- Relevant raft commands 
- Few very naive strategies for migration (just to have something)

karmeli87 committed:
RavenDB-13106 first steps towards resharding

ayende commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -70,6 +71,12 @@ public class ShardRangeAssignment
+
+        public long MigrationCutOffIndex;
+
+        public string ShardedDatabaseId;
```
What is that for?

karmeli87 replied:
this is for the "dbid" part of the `MOVE` tag in the migrated change vector.

ayende commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -70,6 +71,12 @@ public class ShardRangeAssignment
             public int Shard;
         }
 
+        public Dictionary<int, BucketMigration> BucketMigrations;
```
This allows concurrent migrations?

karmeli87 replied:
Yes, having one at a time it is only a matter of policy. 
(currently it is blocked to have concurrent migrations, but we can unblock it)

ayende commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -70,6 +71,12 @@ public class ShardRangeAssignment
 
+        public Dictionary<int, BucketMigration> BucketMigrations;
+
+        public long MigrationCutOffIndex;
```
what is the meaning of this?

karmeli87 replied:
at some point we want to clean the old 'Pointers', so this is the cutoff index that will tell us upon which index to clean.
This will be also used to determine for the change vector when the `MOVE` tag is permanent.

ayende replied:
Okay, make sense. The idea is that we'll have a list of migrations sorted by dates, and we'll use that, right?

ayende replied:
To set the `MigrationCutOffIndex` 

Best to add a comment explaining this

karmeli87 replied:
done

ayende commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -489,6 +496,58 @@ public enum DeletionInProgressStatus
+        Pending,
+        Moving,
+        Moved,
+        OwnershipTransferred
```
Isn't this one just the same as not having a migration at all?

karmeli87 replied:
we need this state, so the source will start the cleaning phase.

ayende replied:
I don't like that, to be honest. That is what `Moved` is for.
So the stages are:
* `BucketMigrations[bucketId].Status = Moving` - start moving data
* `BucketMigrations[bucketId].Status = Moved` - start removing data
* `BucketMigrations[bucketId]` <- missing - meaning that the process is completely done.

Any remaining will be done with the background cleanup

karmeli87 replied:
I want to avoid moving the bucket again if there are still leftovers from the previous move.

ayende replied:
That isn't meant to be here. This is a policy decision. We'll be implementing that after, so for now that is a manual operation.

When we need it, we can have a "history" section that we'll use to remember the recent moves.

karmeli87 replied:
I don't want the user to move bucket 123 from $0 to $1 and allow him to move this bucket again before the cleanup is over

ayende replied:
That is when it is still in `Moved` state

karmeli87 replied:
Since most the part using the state isn't really wired, I will leave as it is for now and we discuss this once more code will be added.

Will add comments to each of the states

ayende commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -489,6 +496,58 @@ public enum DeletionInProgressStatus
+        public int Bucket;
+        public int SourceShard;
+        public int DestinationShard;
+        public long MigrationIndex;
+        public long? ConfirmationIndex;
```
Those are the raft indexes?

karmeli87 replied:
yes

ayende commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -489,6 +496,58 @@ public enum DeletionInProgressStatus
+        public long? ConfirmationIndex;
+        public string LastSourceChangeVector;
+
+        public List<string> ConfirmedDestinations = new List<string>();
+        public List<string> ConfirmedSourceCleanup = new List<string>();
```
Why are those lists?

karmeli87 replied:
- `ConfirmedDestinations` the destination will confirm all its member nodes in order to move the status of the migration to `OwnershipTransffered` and change the buckets mapping.
- `ConfirmedSourceCleanup` the source will confirm the cleanup after the ownership was transferred and only when all source nodes confirmed we will remove this migration from the dictionary

ayende replied:
For cleanup, I would rather that we do that on the source node, as before, then replicate that normally

karmeli87 replied:
This part isn't wired yet. Will discuss this for the next PRs

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -118,6 +118,12 @@ private async Task HandleClusterDatabaseChanged(string databaseName, long index,
+                            if (type == nameof(DestinationMigrationConfirmCommand))
+                            {
+                                if (rawRecord.BucketMigrations.Any(m => m.Value.ConfirmationIndex == index))
+                                    _shardedDatabases.Clear();
```
What is that about? `_shardedDatabases` is a global value, no? Why are you clearing all of it?
Also, what about disposing of `ShardedContext` ?

ayende replied:
What would happen on snapshot install for this?

karmeli87 replied:
this is needed only for the request handlers to update their new buckets mapping after the migration.

ayende replied:
Assume that the mapping update happens on snapshot install.

Also, you are now removing this for _all_ databases.

I _really_ don't like this apparoach

karmeli87 replied:
If I understand correctly we are talking here about 2 things
1. What happened on snapshot install -> I would say clear it all.
2. Upon migration completion -> Yes we need to update only relevant dbs (instead of clearing)

Do I get that right?

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -193,6 +194,19 @@ public DocumentDatabase(string name, RavenConfiguration configuration, ServerSto
             }
         }
 
+        public void SetIds(RawDatabaseRecord record)
+        {
+            DatabaseGroupId = record.Topology.DatabaseTopologyIdBase64;
+            ClusterTransactionId = record.Topology.ClusterTransactionIdBase64;
+            ShardedDatabaseId = record.ShardedDatabaseId;
+        }
+        public void SetIds(DatabaseRecord record)
+        {
+            DatabaseGroupId = record.Topology.DatabaseTopologyIdBase64;
+            ClusterTransactionId = record.Topology.ClusterTransactionIdBase64;
+            ShardedDatabaseId = record.ShardedDatabaseId;
+        }
```
Change this to accept `Topology` ?

ayende commented on src/Raven.Server/Documents/Handlers/ReplicationHandler.cs:
```diff
@@ -305,6 +305,7 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
                     {
                         ["Key"] = new DynamicJsonValue
                         {
+                            [nameof(item.Key)] = item.Key.GetType(),
```
`GetType` ?

karmeli87 replied:
yeah, it can be any type that inherit from the `ReplicationNode` class

karmeli87 replied:
added `ToString()`

ayende commented on src/Raven.Server/Documents/Replication/DocumentInfoHelper.cs:
```diff
@@ -24,6 +24,22 @@ public unsafe LazyStringValue GetDocumentId(Slice key)
             return _tmpLazyStringInstance;
         }
+        
+        public LazyStringValue GetDocumentId(ReplicationBatchItem item)
```
That is code that was moved, no?

karmeli87 replied:
not sure I follow.. this is a new code.

ayende replied:
This is very similar to code that I'm familiar with. Is this code duplication?

karmeli87 replied:
indeed :)
similar to the `ShouldAllow` from the `AllowedPathsValidator` but serve different purpose.
Maybe eventually they will be unified.

ayende replied:
Okay, make a note to clean it later, then

karmeli87 replied:
done

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -947,6 +959,92 @@ private void HandleTopologyChange(DatabaseRecord newRecord)
+            // remove
+            foreach (var handler in OutgoingHandlers)
+            {
+                if (handler is OutgoingMigrationReplicationHandler migrationHandler == false)
```
```suggestion
                if (handler is not OutgoingMigrationReplicationHandler migrationHandler)
```

karmeli87 replied:
done

ayende commented on src/Raven.Client/Documents/Operations/Replication/BucketMigrationReplication.cs:
```diff
@@ -0,0 +1,54 @@
+
+        public bool ForBucketMigration(BucketMigration migration)
+        {
+            if (migration.MigrationIndex != MigrationIndex)
```
Can we have different migration index on same bucket / shard ?

karmeli87 replied:
Yes, but not concurrently.
consider  
 - move bucket 123 from 0 to 1 at index 10
 - move bucket 123 from 1 to 0 at index 20
 - move bucket 123 from 0 to 3 at index 30

first and last has same shard/bucket but different index

ayende replied:
So why do we need the `MigrationIndex` check?

karmeli87 replied:
Precaution, to make sure we handle the correct migration.

ayende commented on src/Raven.Server/Documents/Replication/Senders/MigrationReplicationDocumentSender.cs:
```diff
@@ -0,0 +1,115 @@
+            // TODO : temporary, waiting for RavenDB-17760
+            var migrationId = ctx.DocumentDatabase.ShardedDatabaseId;
+            using var helper = new DocumentInfoHelper();
+            foreach (var item in base.GetReplicationItems(ctx, etag, stats, caseInsensitiveCounters))
```
This does linear scan for now, until we have the bucket index.
Important that we can do that on a per etag as well, so big buckets can be done in batches

ayende commented on src/Raven.Server/Documents/Replication/Senders/MigrationReplicationDocumentSender.cs:
```diff
@@ -0,0 +1,115 @@
+                        return true;
+                    }
+
+                    if (doc.Id.StartsWith(HiLoHandler.RavenHiloIdPrefix, StringComparison.OrdinalIgnoreCase))
```
why are you skipping hilos?

karmeli87 replied:
hmm I was under the impression that the HiLo scope is per shard and not per bucket.. will remove this line.

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -2696,6 +2700,10 @@ private void SqueezeDatabasesToSingleNodeCluster(ClusterOperationContext context
             {
                 using (var rawRecord = ReadRawDatabaseRecord(context, name))
                 {
+                    // we can't squeeze sharded database to a single node
+                    if (rawRecord.IsSharded())
+                        continue;
+
```
huh? That is when we have a single node in the cluster, no?

karmeli87 replied:
that is when we do a hard reset to a single node

ayende replied:
So why aren't we handling this for shards a well?

karmeli87 replied:
for non sharded we will end up with 1 node database or with no database at all.
here we might end up only with some of the shards... let's we will end up only with $1,$4 out of 5 shards.

ayende replied:
Oh, I see. Make sense, then

ayende replied:
Maybe add alert?

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -3429,6 +3437,22 @@ public DatabaseTopology ReadDatabaseTopology<TTransaction>(TransactionOperationC
+                if (databaseRecord.IsSharded() == false)
+                    throw new InvalidOperationException($"The database record '{name}' is not sharded.");
+
+                var topology = databaseRecord.Shards[shard];
```
Check the length here, so we'll get good error and not just index out of range without knowing which db

karmeli87 replied:
done

ayende commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+
+            if (migration.Status != MigrationStatus.Moving)
+                throw new InvalidOperationException($"Expected status is '{MigrationStatus.Moving}', Actual '{migration.Status}'");
+
```
How do you recover from this?

karmeli87 replied:
manually if needed. Note that those exceptions wouldn't jam the state machine. They will be caught and rethrow them as `RachisApplyException`

ayende commented on src/Raven.Server/Utils/ShardHelper.cs:
```diff
@@ -54,5 +132,144 @@ public static IEnumerable<string> GetShardNames(DatabaseRecord record)
+            try
+            {
+                if (bucket >= NumberOfBuckets)
+                    throw new InvalidOperationException($"Total number of buckets is {NumberOfBuckets}, requested: {bucket}");
```
In what db?

karmeli87 replied:
done

ayende reviewed (commented):
Need to spend more time on `BucketsMigrator` review, but here is my initial one

karmeli87 committed:
RavenDB-13106 address PR comments

ayende commented on src/Raven.Server/Documents/Handlers/ReplicationHandler.cs:
```diff
@@ -305,7 +305,7 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
                         ["Key"] = new DynamicJsonValue
                         {
-                            [nameof(item.Key)] = item.Key.GetType(),
+                            [nameof(item.Key)] = item.Key.GetType().ToString(),
```
That is the type, so why is it called `Key`.
Also, let's avoid adding type names to the protocol. Let's use an enum

karmeli87 replied:
this isn't a protocol.. this a debug/studio EP
This also the reason why I stayed with the "Key" here.

ayende replied:
That is part of our endpoints, which is why I would rather have an enum over type name, which may change arbitrarily

karmeli87 committed:
RavenDB-13106 address PR comments

make a distiguish between 'DatabasesCache' which contain only phisical dbs e.g. NW$0, NW$1, ...
and 'shardedDatabases' which contain virtual/sharded db e.g. NW

ppekrol commented on src/Raven.Client/Documents/Operations/Replication/BucketMigrationReplication.cs:
```diff
@@ -0,0 +1,54 @@
+using Sparrow;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Client.Documents.Operations.Replication
```
move to a dedicated .Sharding namespace?

ppekrol commented on src/Raven.Client/Documents/Operations/Replication/BucketMigrationReplication.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace Raven.Client.Documents.Operations.Replication
+{
+    public class BucketMigrationReplication : ReplicationNode
```
internal?

ppekrol commented on src/Raven.Client/Documents/Operations/Replication/BucketMigrationReplication.cs:
```diff
@@ -0,0 +1,54 @@
+    public class BucketMigrationReplication : ReplicationNode
+    {
+        public int Bucket;
+        public int Shard;
```
better to set those to via ctor and validate their values?

ppekrol commented on src/Raven.Client/Documents/Operations/Replication/BucketMigrationReplication.cs:
```diff
@@ -0,0 +1,54 @@
+            MigrationIndex = migrationIndex;
+        }
+
+        public bool ForBucketMigration(BucketMigration migration)
```
static? not sure how this is bound to this class?

karmeli87 replied:
this methods check whether the an incoming bucket migration (the `BucketMigration migration` input) is already handled by this class.

ppekrol commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -70,6 +71,15 @@ public class ShardRangeAssignment
             public int Shard;
         }
 
+        public Dictionary<int, BucketMigration> BucketMigrations;
```
Is there a possibility to wrap all of those into some .Sharding property?

ppekrol commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -489,6 +499,68 @@ public enum DeletionInProgressStatus
         HardDelete
     }
 
+    public enum MigrationStatus
```
move at least that to partial

ppekrol commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -489,6 +499,68 @@ public enum DeletionInProgressStatus
+        OwnershipTransferred
+    }
+
+    public class BucketMigration : IDatabaseTask
```
separate file, separate namespace?

ppekrol replied:
isn't this a server-side thing?

karmeli87 replied:
this is part of the database record

ppekrol commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -193,6 +194,19 @@ public DocumentDatabase(string name, RavenConfiguration configuration, ServerSto
+        {
+            DatabaseGroupId = record.Topology.DatabaseTopologyIdBase64;
+            ClusterTransactionId = record.Topology.ClusterTransactionIdBase64;
+            ShardedDatabaseId = record.ShardedDatabaseId;
```
create private method that accepts topology and shardedDatabaseId, use it in both places or just remove those and use new method

ppekrol commented on src/Raven.Server/Documents/Replication/Incoming/IncomingMigrationReplicationHandler.cs:
```diff
@@ -0,0 +1,39 @@
+using Sparrow.Json;
+using Sparrow.Server;
+
+namespace Raven.Server.Documents.Replication.Incoming
```
.Sharding

ppekrol commented on src/Raven.Server/Documents/Replication/Outgoing/OutgoingMigrationReplicationHandler.cs:
```diff
@@ -0,0 +1,53 @@
+using Raven.Server.Utils;
+using Sparrow.Logging;
+
+namespace Raven.Server.Documents.Replication.Outgoing
```
.Sharding

ppekrol commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -947,6 +959,92 @@ private void HandleTopologyChange(DatabaseRecord newRecord)
+            using (_server.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx))
+            using (ctx.OpenReadTransaction())
+            {
+                return _server.Cluster.ReadDatabaseTopologyForShard(ctx, ShardHelper.ToDatabaseName(Database.Name), shard);
```
Cluster.Sharding.

ppekrol commented on src/Raven.Server/Documents/Replication/Senders/MigrationReplicationDocumentSender.cs:
```diff
@@ -0,0 +1,108 @@
+using Sparrow.Server;
+using Sparrow.Server.Utils;
+
+namespace Raven.Server.Documents.Replication.Senders
```
.Sharding

ppekrol commented on src/Raven.Server/Documents/Sharding/ShardedContext.cs:
```diff
@@ -14,11 +16,11 @@ public unsafe class ShardedContext
+        public ShardedContext(ServerStore server, RawDatabaseRecord record)
         {
             //TODO: reduce the record to the needed fields
             _record = record;
```
you are converting that Raw to materialized anyway here

karmeli87 replied:
I know... this needs to be handled properly and take only relevant fields. 
change the `TODO` to `TODO: [sharding]`

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -3437,6 +3445,26 @@ public DatabaseTopology ReadDatabaseTopology<TTransaction>(TransactionOperationC
             }
         }
 
+        public DatabaseTopology ReadDatabaseTopologyForShard<TTransaction>(TransactionOperationContext<TTransaction> context, string name, int shard)
```
move that to .Sharding property and handle this there

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+using Raven.Server.Utils;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.ServerWide.Commands
```
.Sharding

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+    {
+        public int SourceShard;
+        public int DestinationShard;
+        public int Bucket;
```
main ctor should set those and validate them

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+{
+    public class StartBucketMigrationCommand : UpdateDatabaseCommand
+    {
+        public int SourceShard;
```
readonly? private set?

karmeli87 replied:
IIRC it can't be readonly due to the de-serializer

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    public class SourceMigrationSendCompletedCommand : UpdateDatabaseCommand
```
separate file?

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+
+    public class SourceMigrationSendCompletedCommand : UpdateDatabaseCommand
+    {
+        public int Bucket;
```
main ctor should validate and set them

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    public class DestinationMigrationConfirmCommand : UpdateDatabaseCommand
```
separate file

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+    public class DestinationMigrationConfirmCommand : UpdateDatabaseCommand
+    {
+        public int Bucket;
+        public long MigrationIndex;
```
same

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    public class SourceMigrationCleanupCommand : UpdateDatabaseCommand
```
separate file

ppekrol commented on src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs:
```diff
@@ -0,0 +1,181 @@
+
+    public class SourceMigrationCleanupCommand : UpdateDatabaseCommand
+    {
+        public int Bucket;
```
same

ppekrol commented on src/Raven.Server/ServerWide/Maintenance/BucketsMigrator.cs:
```diff
@@ -0,0 +1,291 @@
+using Raven.Server.Utils;
+using Sparrow;
+
+namespace Raven.Server.ServerWide.Maintenance
```
.Sharding.Migration

ppekrol commented on src/Raven.Server/ServerWide/Maintenance/BucketsMigrator.cs:
```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    public class MergedDatabaseStatusReport
```
separate file?

ppekrol commented on src/Raven.Server/ServerWide/Maintenance/BucketsMigrator.cs:
```diff
@@ -0,0 +1,291 @@
+        }).ToArray();
+    }
+
+    public class ShardMigrationResult
```
separate file?

ppekrol commented on src/Raven.Server/ServerWide/Maintenance/BucketsMigrator.cs:
```diff
@@ -0,0 +1,291 @@
+
+        public override int GetHashCode() => (int)Hashing.XXHash64.Calculate(new int[] { SourceShard, DestinationShard, Bucket });
+
+        public override bool Equals(object? obj)
```
remove '?'

ppekrol commented on src/Raven.Server/ServerWide/Maintenance/BucketsMigrator.cs:
```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    public class ShardReport
```
move to separate files

karmeli87 committed:
RavenDB-13106 address PR comments
